### PR TITLE
Storage rewrite

### PR DESCRIPTION
**Pull request by atk**
The storage package was needlessly complex, large and led to frequent misunderstandings from developers who expected it to work like a store. I have learned from my mistakes and deprecated the old APIs with the exception of the cookieStorage. I also added a new smaller `makePersisted(signalOrStore)` primitive that should fulfil the general requirements and improved the cookieStorage.

### CODE REVIEW

File: configs/vitest.config.ts [hunk 1]
```diff
@@ -45,6 +45,12 @@ export default defineConfig(({ mode }) => {
     },
     resolve: {
       conditions: testSSR ? ["node"] : ["browser", "development"],
+      // fix issues with stores and node@20:
+      ...(!testSSR && { 
+        alias: {
+          "solid-js/store": "solid-js/store/dist/store.js"
```
Comment (thetarnav) [inline configs/vitest.config.ts:51]: Could you explain this? do we need the same thing for `"solid-js/web"` too? Also we have set the required node version to `18` for the repo. Maybe we should add a [`.nvmrc`](https://github.com/thetarnav/solid-devtools/blob/main/.nvmrc) file to enforce it?

Comment (atk) [reply configs/vitest.config.ts:51]: Node@20 partially messes up the resolver for vitest. I have not yet found the reasons, only two workarounds: make solid an external dependency or alias some files.

**Review by thetarnav — APPROVED** [review]
I like this change, it's going in the right direction with simplicity and composability.
Although I would split the primitive into multiple parts, instead of simply wrapping a store/signal tuple.
What if I have a custom state API that requires special reconciliation, or I'm using a store and want to reconcile manually with `produce`?
What if my state API does not provide a tuple?
What if I want to use it with `createResource`?
Like it would be nice to split the API into 'serialization', 'deserialization', 'getting initial value', 'updating storage on state change' and 'updating state on storage change'. This way the user would have more clarity into what the primitive will actually do, and how he can customize it, instead of just configuring the options object to some limited extend.

**Comment by atk** [comment]
> What if I have a custom state API that requires special reconciliation, or I'm using a store and want to reconcile manually with produce? What if my state API does not provide a tuple?

Then you're welcome to write an adapter that projects your custom state API to something resembling either a signal or a store.
I want the simplest possible API for the default use cases, though I'm happy with splitting this like you suggested, as long as the main API remains the same. I'd prefer to get this out in any case. So I'd prefer to have this merged now and take some time to split this up afterwards.

> What if I want to use it with createResource?

Then you should use `createResource(fetcher, { storage: makePersistent(createSignal()) })`; anything else might make things unneccessary complex. But you're right, I should document that.

**Comment by thetarnav** [comment]
> I'd prefer to get this out in any case. So I'd prefer to have this merged now and take some time to split this up afterwards.

yup go ahead, just wanted to get my thoughts out. The new API is definitely an improvement.

> I should document that.

💯😊

**Pull request merged**